opentracing-agent/src/main/java/io/opentracing/contrib/agent/OpenTracingHelper.java

```diff
@@ -0,0 +1,285 @@
+import java.util.logging.Logger;
+
+import org.jboss.byteman.rule.Rule;
+import org.jboss.byteman.rule.helper.Helper;
+
+import io.opentracing.Span;
+import io.opentracing.SpanContext;
+import io.opentracing.Tracer;
+import io.opentracing.Tracer.SpanBuilder;
+import io.opentracing.contrib.global.GlobalTracer;
+import io.opentracing.contrib.spanmanager.DefaultSpanManager;
+import io.opentracing.contrib.spanmanager.SpanManager;
+import io.opentracing.contrib.spanmanager.SpanManager.ManagedSpan;
+import io.opentracing.propagation.Format;
+
+/**
+ * This class provides helper capabilities to the byteman rules.
+ */
+public class OpenTracingHelper extends Helper {
+
+    private static final Logger log = Logger.getLogger(OpenTracingHelper.class.getName());
+
+    private static Tracer tracer = new AgentTracer(GlobalTracer.get());
+    private static SpanManager spanManager = DefaultSpanManager.getInstance();
+
+    private static Map<Object,Span> spanAssociations = Collections.synchronizedMap(new WeakHashMap<Object,Span>());
+    private static Map<Object,Span> finished = Collections.synchronizedMap(new WeakHashMap<Object,Span>());
+
+    private static Map<Object,Integer> state = Collections.synchronizedMap(new WeakHashMap<Object,Integer>());
+
+    public OpenTracingHelper(Rule rule) {
+        super(rule);
+    }
+
+    /**
+     * This method returns the OpenTracing tracer.
+     *
+     * @return The tracer
+     */
+    public Tracer getTracer() {
+        return tracer;
+    }
+
+    /**
+     * This method requests that the supplied span becomes the
+     * current 'active' span.
+     *
+     * @param span The span
+     */
+    public void activateSpan(Span span) {
+        spanManager.activate(span);
+    }
+
+    /**
+     * This method returns the currently active span.
+     *
+     * @return The current span, or null if no active span exists
+     */
+    public Span currentSpan() {
+        return spanManager.current().getSpan();
+    }
+
+    /**
+     * This method requests that the current active span
+     * should be deactivated. If the underlying
+     * span management mechanism maintains a stack, then
+     * this will result in the previous (unfinished) parent
+     * span being reinstated as the current active span.
+     *
+     * @return The span being deactivated, or null if no active span found
+     */
+    public Span deactivateCurrentSpan() {
+        ManagedSpan current = spanManager.current();
+        current.deactivate();
+        return current.getSpan();
+    }
+
+    /**
+     * This method establishes an association between an application object
+     * (i.e. the subject of the instrumentation) and a span. Once the
+     * application object is no longer being used, the association with the
+     * span will automatically be discarded.
+     *
+     * @param obj The application object to be associated with the span
+     * @param span The span
+     */
+    public void associateSpan(Object obj, Span span) {
```

Comment:
It is used in the `java-net.btm` rules, to associate the span with the `HttpURLConnection` instance, and then `retrieveSpan(Object obj)` is used to retrieve the span for the connection in subsequent rules. It is for cases where the target object being instrumented does not have a way to store context itself.